src/foam/mlang/mlang.js

```diff
@@ -4044,6 +4044,124 @@ foam.CLASS({
   ]
 });
 
+foam.CLASS({
+  package: 'foam.mlang.predicate',
+  name: 'NamedProperty',
+  extends: 'foam.mlang.AbstractExpr',
+  implements: [ 'foam.core.Serializable' ],
+
+  documentation: `Stores propName as a property and returns property when f() is called.`,
+
+  javaImports: [
+    'foam.core.ClassInfo',
+    'foam.core.FObject',
+    'foam.core.PropertyInfo',
+    'foam.mlang.Expr',
+    'java.util.List',
+    'java.util.Map',
+    'java.util.concurrent.ConcurrentHashMap'
+  ],
+
+  axioms: [
+    foam.pattern.Multiton.create({property: 'propName'}),
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(
+          `
+  protected final static Map map__ = new ConcurrentHashMap();
+  public static NamedProperty create(String propName) {
+    NamedProperty p = (NamedProperty) map__.get(propName);
+
+    if ( p == null ) {
+      p = new NamedProperty();
+      p.setPropName(propName);
+      map__.put(propName, p);
+    }
+
+    return p;
+  }
+ `
+        );
+      }
+    }
+  ],
+
+  properties: [
+    {
+      class: 'Map',
+      name: 'specializations_',
```

Comment:
I'm not sure this is worthwhile since now we just look it up in the specialization map instead of the axiom map, which should be the same.